README.md
# React Google Publisher Tag

## Motivation

I needed to show adsense via google publisher tag in my react isomorphic application.
Please read carefully (AdSense terms)[https://support.google.com/adsense/answer/48182?hl=en].
You are using this module on your own risk.

## Install
```sh
npm install react-google-publisher-tag
```

## Features

 * Automatically show AD and fit the size
 * Automatically load google script
 * Support for mobile (you can show different size for mobile users)
 * Support for isomorphic applications
 * Support for responsible layout


## Usage

 - Go to https://www.google.com/dfp and create a new Ad unit. (tab Inventory)
 - Add sizes: 120x240, 120x600, 125x125, 160x600, 180x150, 200x200, 234x60, 240x400, 250x250, 300x100, 300x250, 300x600, 320x50, 336x280, 468x60, 728x90, 970x90
 - Use ad code as path in the react component


### App.jsx

Application part (load google analytics script to your webpage on the client side).
ReactGAnalytics has parameter ID (use your own ID)

```js
var React = require('react');
var AD = require('react-google-publisher-tag');

var App = module.exports = React.createClass({
  render: function() {
    return (
      <div id="application">
        <AD path="/network-code/ad-code" />
        <RouteHandler />
      </div>
    );
  }
});
```

## Parameters

| Name           |     Type  |  Required | Default    |
|----------------|-----------|-----------|------------|
| path           | String    | true      |            |
| responsive     | Boolean   | true      | true       |
| format         | String    | true      | HORIZONTAL |
| canBeLower     | Boolean   | true      | true       |
| dimensions     | Array     | false     |            |
| minWindowWidth | Integer   | false     | -1         |
| maxWindowWidth | Integer   | false     | -1         |
| className      | String    | false     |            |

## Path

Path is only one required parameter. The format is:
/network-code/ad-code

Example:

```js
/22222222/myad
```

## Network code

You can find your network code in the Admin tab (Google DFP).
This unique code is included in DFP tags to identify the network, ensuring that DFP looks in the correct network for line items that are targeted to a particular ad unit. It's displayed here to help you identify which network is associated with any DFP tags you have on your site.

## Ad code

You can find your ad code when you click on your ad in the Google DFP (Inventary tab)
Identifies the ad unit in the associated ad tag. Ad unit codes can be up to 100 characters in length. Only letters, numbers, underscores, hyphens, periods, asterisks, forward slashes, backslashes, exclamations, left angle brackets, colons and parentheses are allowed. Each code must be unique; you can't reuse codes you've used before. Once you've created the ad unit, you can't change the code.


## Formats

It will setup ad size automatically for this component:
 - HORIZONTAL [[970, 90], [728, 90], [468, 60], [234, 60]]
 - RECTANGLE [[336, 280], [300, 250], [250, 250], [200, 200], [180, 150], [125, 125]]
 - VERTICAL [[300, 600], [160, 600], [120, 600], [120, 240]]
 - MOBILE: [[320, 50]],
 - 300x600: [[300, 600], [160, 600]]
 - 336x280: [[336, 280], [300, 250]]
 - 728x90: [[728, 90], [468, 60]]
 - 970x90: [[970, 90], [728, 90], [468, 60]]

## Can be lower

You can allow lower ad size automatically. (Default: true)

## Credits

[Zlatko Fedor](http://github.com/seeden)

## License

The MIT License (MIT)

Copyright (c) 2015 Zlatko Fedor zlatkofedor@cherrysro.com


History.md
# Change/History Log

This project adheres to [Semantic Versioning](http://semver.org/).
Every release, along with the migration instructions, is documented on the Github [Releases](https://github.com/seeden/react-google-publisher-tag/releases) page.


src/index.js
/**
 * https://developers.google.com/doubleclick-gpt/reference
*/
import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import keymirror from 'keymirror';

export const Format = keymirror({
  HORIZONTAL: null,
  RECTANGLE: null,
  VERTICAL: null,
  MOBILE: null,
});

export const Dimensions = {
  [Format.HORIZONTAL]: [[970, 90], [728, 90], [468, 60], [234, 60]],
  [Format.RECTANGLE]: [[336, 280], [300, 250], [250, 250], [200, 200], [180, 150], [125, 125]],
  [Format.VERTICAL]: [[300, 600], [160, 600], [120, 600], [120, 240]],
  [Format.MOBILE]: [[320, 50]],
  '300x600': [[300, 600], [160, 600]],
  '336x280': [[336, 280], [300, 250]],
  '728x90': [[728, 90], [468, 60]],
  '970x90': [[970, 90], [728, 90], [468, 60]],
};

function prepareDimensions(dimensions, format = Format.HORIZONTAL, canBeLower = true) {
  if (!dimensions || !dimensions.length) {
    return Dimensions[format] || [];
  }

  if (dimensions.length === 1 && canBeLower) {
    const dimension = dimensions[0];
    const key = `${dimension[0]}x${dimension[1]}`;

    if (Dimensions[key]) {
      return Dimensions[key] || [];
    }
  }

  return dimensions;
}

let nextID = 1;
let googletag = null;

function getNextID() {
  return 'rgpt-' + (nextID++);
}

function initGooglePublisherTag() {
  if (googletag) {
    return;
  }

  googletag = window.googletag = window.googletag || {};
  googletag.cmd = googletag.cmd || [];

  googletag.cmd.push(function prepareGoogleTag() {
    // add support for async loading
    googletag.pubads().enableAsyncRendering();

    // collapse div without ad
    googletag.pubads().collapseEmptyDivs();

    // load ad with slot refresh
    googletag.pubads().disableInitialLoad();

    // enable google publisher tag
    googletag.enableServices();
  });

  (function loadScript() {
    const gads = document.createElement('script');
    gads.async = true;
    gads.type = 'text/javascript';
    gads.src = '//www.googletagservices.com/tag/js/gpt.js';

    const head = document.getElementsByTagName('head')[0];
    head.appendChild(gads);
  })();
}

export default class GooglePublisherTag extends Component {
  static propTypes = {
    className: React.PropTypes.string,
    path: React.PropTypes.string.isRequired,
    format: React.PropTypes.string.isRequired,
    responsive: React.PropTypes.bool.isRequired,
    canBeLower: React.PropTypes.bool.isRequired, // can be ad lower than original size,

    dimensions: React.PropTypes.array,  // [[300, 600], [160, 600]]

    minWindowWidth: React.PropTypes.number.isRequired,
    maxWindowWidth: React.PropTypes.number.isRequired,
  };

  static defaultProps = {
    format: Format.HORIZONTAL,
    responsive: true,
    canBeLower: true,
    dimensions: null,
    minWindowWidth: -1,
    maxWindowWidth: -1,
  };

  componentDidMount() {
    initGooglePublisherTag();

    if (this.props.responsive) {
      window.addEventListener('resize', this.handleResize);
    }

    googletag.cmd.push(() => {
      this.initialized = true;

      this.update(this.props);
    });
  }

  componentWillReceiveProps(props) {
    this.update(props);
  }

  componentWillUnmount() {
    // TODO sometimes can props changed
    if (this.props.responsive) {
      window.removeEventListener('resize', this.handleResize);
    }

    this.removeSlot();
  }

  update(props) {
    if (!this.initialized) {
      return;
    }

    const node = findDOMNode(this);
    if (!node) {
      return;
    }

    const componentWidth = node.offsetWidth;
    const availableDimensions = prepareDimensions(props.dimensions, props.format, props.canBeLower);

    // filter by available node space
    let dimensions = props.responsive
      ? availableDimensions.filter((dimension) => dimension[0] <= componentWidth)
      : availableDimensions;


    // filter by min and max width
    const windowWidth = window.innerWidth;
    const { minWindowWidth, maxWindowWidth } = props;

    if (minWindowWidth !== -1 && minWindowWidth < windowWidth) {
      dimensions = [];
    } else if (maxWindowWidth !== -1 && maxWindowWidth > windowWidth) {
      dimensions = [];
    }

    // do nothink
    if (JSON.stringify(dimensions) === JSON.stringify(this.currentDimensions)) {
      return;
    }

    this.currentDimensions = dimensions;


    if (this.slot) {
      // remove current slot because dimensions is changed and current slot is old
      this.removeSlot();
    }

    // there is nothink to display
    if (!dimensions || !dimensions.length) {
      return;
    }

    if (!this.refs.holder) {
      console.log('RGPT holder is undefined');
      return;
    }

    // prepare new node
    const id = getNextID();
    this.refs.holder.innerHTML = `<div id="${id}"></div>`;

    // prepare new slot
    const slot = this.slot = googletag.defineSlot(props.path, dimensions, id);
    slot.addService(googletag.pubads());

    // display new slot
    googletag.display(id);
    googletag.pubads().refresh([slot]);
  }

  removeSlot() {
    if (!this.slot) {
      return;
    }

    googletag.pubads().clear([this.slot]);
    this.slot = null;

    if (this.refs.holder) {
      this.refs.holder.innerHTML = null;
    }
  }

  refreshSlot() {
    if (this.slot) {
      googletag.pubads().refresh([this.slot]);
    }
  }

  handleResize = () => {
    this.update(this.props);
  }

  render() {
    return (
      <div className={this.props.className} ref="holder"></div>
    );
  }
}
